### PR TITLE
ContextualMenu: Make sure vertical scrollbar does not take space away from contents

### PR DESCRIPTION
**Pull request by christiango**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ npm run change`

#### Description of changes
We currently have inconsistent behavior across browsers when a contextual menu is positioned such that a vertical scrollbar has to appear. This scenario usually occurs when a menu is launched below the target with directionalHintFixed set to true when the browser window is not very tall or the menu is very long.

Chrome places the scrollbar outside the contents of the menu, which means the scrollbar does not take space away from the contents:
image

Other browsers instead take space away from the content to position the scrollbar. In the event that there are long menu items, this results in the content of menu items being truncated:
Edge:
image
 
FireFox:
image 

IE 11:
image

After experimenting a bit and doing some research, it appears that setting `overflow-y:scroll` for all browsers causes them to place the scrollbar outside the container. However, this means that the scrollbar is actually visible, even when the contents fit. In order to make it such that scrollbars appear without truncating content, we need to manually determine if we need to show a scrollbar and set the overflow property directly.

This is the results in Edge after the fix:
image

#### Focus areas to test
Since we're doing measurements I made sure this didn't introduce a reflow using the dev tools in Edge, FF, and Chrome. All browsers showed that there was already a reflow from CalloutContent._getMaxHeight() and the additional measurements for this bug fix do not introduce an additional reflow.

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/Callout/CalloutContent.tsx [hunk 1]
```diff
@@ -146,6 +147,12 @@ export class CalloutContent extends BaseComponent<ICalloutProps, ICalloutState>
     let getContentMaxHeight: number = this._getMaxHeight() + this.state.heightOffset!;
     let contentMaxHeight: number = calloutMaxHeight! && (calloutMaxHeight! < getContentMaxHeight) ? calloutMaxHeight! : getContentMaxHeight!;
 
+    // When using overflow-y: auto, most browsers take away room from the content div to place the scrollbar. This can result
+    // in truncated content, since there isn't as much horizontal room available. If overflow-y:scroll is set, then the scrollbar
+    // does not take away from the content width. Overflow-y: scroll always shows a scrollbar, so we need to do some measurements
+    // to check if we actually need a scrollbar.
+    const needsVerticalScrollbar: boolean = !!(this._popup && this._popup.refs.root && this._popup.refs.root.scrollHeight > this._popup.refs.root.clientHeight);
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Callout/CalloutContent.tsx]: hmmm.. this seems really gross..

Comment (christiango) [reply packages/office-ui-fabric-react/src/components/Callout/CalloutContent.tsx]: Any ideas? I couldn't think of a way to do this in css

Comment (christiango) [reply packages/office-ui-fabric-react/src/components/Callout/CalloutContent.tsx]: I don't like it either, but the bug is pretty bad...

Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Callout/CalloutContent.tsx]: add a method or property to `IPopup` which can evaluate if a vertical scrollbar is needed, OR expose the scrollHeight/clientHeight to be read. Just don't reach into refs. "refs" is going away and really should never have been a public contract. As soon as popup is wrapped in a decorator, this won't work.

**Comment by christiango** [comment]
@dzearing, I avoided setting refs and did the work inside Popout

File: packages/office-ui-fabric-react/src/components/Popup/Popup.tsx [hunk 2]
```diff
@@ -55,7 +55,12 @@ export class Popup extends BaseComponent<IPopupProps, {}> {
   }
 
   public render() {
-    let { role, className, ariaLabel, ariaLabelledBy, ariaDescribedBy } = this.props;
+    let { role, className, ariaLabel, ariaLabelledBy, ariaDescribedBy, style } = this.props;
+
+    let needsVerticalScrollBar = false;
+    if (this.refs.root) {
```
Comment (christiango) [inline packages/office-ui-fabric-react/src/components/Popup/Popup.tsx]: Right now this is leveraging the fact that popup is rendered by a callout using multiple passes, so we'd have a valid root on the second render. This is the same thing that calloutcontent uses for positioning and it gives us the perf we want by not introducing a reflow or another render (like we would get if we did this in componentdidupdate),  but it does feel a bit weird.

File: packages/office-ui-fabric-react/src/components/Popup/Popup.types.ts [hunk 3]
```diff
@@ -50,4 +50,11 @@ export interface IPopupProps extends React.HTMLAttributes<Popup> {
    * @default true
    */
   shouldRestoreFocus?: boolean;
+
+  /**
+   * If set to true, the popup will explicitly check if a scroll bar is needed and set overflow-y to scroll if necessary. This is useful
+   * when you don't want the scroll bar to take away space from the popout content, which is what happens when you use overflow-y:auto
+   * on some browsers.
+   */
+  shouldSetOverflowY?: boolean;
```
Comment (phkuo) [inline packages/office-ui-fabric-react/src/components/Popup/Popup.types.ts]: you're not actually using this param anywhere in code

Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Popup/Popup.types.ts]: hmm yes can we remove this?

Comment (christiango) [reply packages/office-ui-fabric-react/src/components/Popup/Popup.types.ts]: Remove or make it read It? I meant to make this optional behavior, so I was going to make the prop work

**Comment by christiango** [comment]
Removed prop, clearing feedback

**Pull request merged**